annotation/tests/test_tasks_services.py

```diff
@@ -580,8 +789,1416 @@ def test_filter_tasks_db_no_files_or_jobs(
+
+
+@pytest.mark.parametrize(
+    ("items", "expected"),
+    (
+        (
+            [
+                [{"a": 1, "b": 2, "c": 3}, {"d": 4}, {"e": 5}],
+                [{"d": 4}, {"f": 6}],
+                [{"d": 4}, {"g": 7}],
+            ],
+            [{"d": 4}],
+        ),
+        (
+            [
+                [{"a": 1}],
+                [],
+                [{"b": 2}],
+            ],
+            [],
+        ),
+        (
+            [
+                [{"a": 1}, {"b": 2}],
+            ],
+            [{"a": 1}, {"b": 2}],
+        ),
+        ([], []),
+    ),
+)
+def test_get_common_values(items: List[List[Any]], expected: List[Any]):
+    assert services.get_common_values(items) == expected
+
+
+def test_change_ids():
+    all_tasks_objs = {
+        "task1": {"objects": [{"id": 1}, {"id": 2}]},
+        "task2": {"objects": [{"id": 3}, {"id": 4}]},
+    }
+
+    doc_objs = {
+        "doc1": {
+            "objects": [
+                {"id": 1, "links": [2], "children": [3]},
+                {"id": 2},
+                {"id": 3},
+            ]
+        }
+    }
+
+    id_mapping = {(1,): 101, (2,): 102, (3,): 103}
+    mock_get_new_id = MagicMock(
+        side_effect=lambda old_id, _: id_mapping.get((old_id,), old_id)
+    )
+    mock_get_common_ids = MagicMock(
+        side_effect=lambda old_id, _: [id_mapping.get((old_id,), old_id)]
+    )
+    mock_get_links_and_children = MagicMock(
+        side_effect=lambda ids, _, __: (
+            [id_mapping.get((i,), i) for i in ids],
+            [],
+        )
+    )
+    mock_get_common_values = MagicMock(side_effect=lambda objs: objs)
+    with patch("annotation.tasks.services.get_new_id", mock_get_new_id), patch(
+        "annotation.tasks.services.get_common_ids", mock_get_common_ids
+    ), patch(
+        "annotation.tasks.services.get_links_and_children",
+        mock_get_links_and_children,
+    ), patch(
+        "annotation.tasks.services.get_common_values", mock_get_common_values
+    ):
+        services.change_ids(all_tasks_objs, doc_objs, id_mapping)
+
+    assert doc_objs["doc1"]["objects"][0]["id"] == 101
+    assert doc_objs["doc1"]["objects"][0]["links"] == [101]
+    assert doc_objs["doc1"]["objects"][0]["children"] == [3]
+    assert doc_objs["doc1"]["objects"][1]["id"] == 102
+    assert doc_objs["doc1"]["objects"][2]["id"] == 103
+    mock_get_new_id.assert_called_with(*(3, {(1,): 101, (2,): 102, (3,): 103}))
+    mock_get_common_ids.assert_called_with(
+        *(1, {(1,): 101, (2,): 102, (3,): 103})
+    )
+    mock_get_links_and_children.assert_called_with(
+        *(
+            [101],
+            {
+                "task1": {"objects": [{"id": 1}, {"id": 2}]},
+                "task2": {"objects": [{"id": 3}, {"id": 4}]},
+            },
+            {(1,): 101, (2,): 102, (3,): 103},
+        )
+    )
+    mock_get_common_values.assert_called_with([101])
+
+
+def test_change_ids_with_children():
```

Comment:
this test seems to be similar to the previous one, any way we could optimize it?

Reply:
I'll try.